### PR TITLE
Update RuboCop to 0.50.x

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

**Comment by sodabrew** [comment]
I've awoken a slumbering giant.

**Comment by tamird** [comment]
😘

**Comment by tamird** [comment]
I've cherry picked the rspec matcher change from #747 since the alternative (replacing parallel assignment with serial assignment) was causing failures on 1.8.7 and segfaults on REE o.O

**Comment by tamird** [comment]
Looks like there are some (unrelated?) failures; segfault on REE and memory corruption weirdness in some 2.0.0 tests.

**Comment by tamird** [comment]
@sodabrew can i bribe you to take a look at this?

File: examples/threaded.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 # Should never exceed worst case 3.5 secs across all 20 threads
 Timeout.timeout(3.5) do
-  20.times.map do
+  Array.new(20) do
```
Comment (sodabrew) [inline examples/threaded.rb]: There's a style against Integer#times now?

Comment (tamird) [inline examples/threaded.rb]: yep, this is apparently more performant

**Comment by jaredbeck** [comment]
Oh, I didn't realize you were working on this, Tamir.

I opened https://github.com/brianmario/mysql2/pull/909, which upgrades rubocop to 0.41.2. But, it looks like you've done a lot of work here, and I don't want to cause a conflict, so maybe we should close mine.

**Comment by tamird** [comment]
Yeah, this one goes a bit further than yours. Let's see what @sodabrew says.

File: Rakefile [hunk 2]
```diff
@@ -8,17 +9,11 @@ load 'tasks/compile.rake'
 load 'tasks/generate.rake'
 load 'tasks/benchmarks.rake'
 
-# TODO: remove engine check when rubinius stops crashing on RuboCop
```
Comment (sodabrew) [inline Rakefile]: Oh nice, it stopped crashing on recent Rubinius?

Comment (tamird) [reply Rakefile]: I haven't actually tested that yet (I'm about to), but Rubinius support looks have been removed some time ago (1df20eb4aa8666ceea664c01e76eac402332f039).

Comment (tamird) [reply Rakefile]: Rubinius is in a bit of a rough state it seems:
https://github.com/bundler/bundler/issues/6163
https://github.com/rubinius/rubinius/issues/3769

I'll check again when either of those issues is fixed.

**Comment by sodabrew** [comment]
This looks good. Also, I'm thinking about when to drop Ruby 1.9, too.

File: Rakefile (hunk 2 above)
Comment (sodabrew) [reply Rakefile]: Yes, true - Rubinius was in rough shape, and even as an allowed failure in Travis, was just gumming up the works to try to keep up.

**Comment by tamird** [comment]
Great, I'd be happy to prepare a PR that does that (with no urgency to merge) after this goes in.

**Comment by jaredbeck** [comment]
```ruby
gem 'rubocop', '~> 0.50.0' unless RUBY_VERSION =~ /1.9/
```

This is an interesting way to get us to a modern rubocop version! Works for me. I can't think of any downside to skip linting on 1.9.3. There aren't any 1.9.3-specific linter offenses ("cops"), are there?

**Comment by tamird** [comment]
@jaredbeck the linter isn't being run on 1.9.3, but it is still configured to target 1.9 - see `.rubocop.yml`.

0.51 removed the ability to **target** 1.9, which is why this is the terminal version until 1.9 support is dropped from mysql2.

**Comment by tamird** [comment]
@sodabrew thanks for the review. Is this waiting on anything?

**Comment by sodabrew** [comment]
Right - that's what I mean - for the next mysql2 rev, I'm considering whether to drop 1.9 as well, and make Ruby 2.0 the lowest supported version. Given that each version branch has multi year lifespan, that 0.4.x supports all the way back to 1.8.7 gives a lot of breathing room to make 0.5.x more aggressively "modern".

**Comment by tamird** [comment]
Understood. Still, it would be beneficial to merge this as-is and pursue (and evaluate the benefits of) dropping 1.9 in a separate PR -- if nothing else, it will make review at least somewhat palatable.

**Comment by jaredbeck** [comment]
> @sodabrew sodabrew merged commit 17fe755 into brianmario:master 31 minutes ago

:tada: Nice work, Tamir!

**Pull request merged**